cni-plugin/linkerd-cni-validator/src/main.rs

```diff
@@ -0,0 +1,174 @@
+                error!(error = %e, "Failed to validate outbound routing configuration");
+            } else {
+                error!("Failed to validate due to server terminating early");
+            }
+
+            exit(UNSUCCESSFUL_EXIT_CODE);
+        }
+    }
+}
+
+#[tracing::instrument(
+    level = "debug",
+    skip(ready_rx),
+    fields(original_dst = %target_addr),
+    )]
+async fn validate_outbound_redirect(
+    target_addr: SocketAddr,
+    ready_rx: oneshot::Receiver<Result<()>>,
+) -> Result<()> {
+    let timeout = std::time::Duration::from_secs(120);
+    if tokio::time::timeout(timeout, ready_rx).await.is_err() {
+        anyhow::bail!("timed-out ({:?}) waiting for server to be ready", timeout);
+    }
+
+    let mut stream = {
+        debug!("Building validation client");
+        let socket = TcpStream::connect(target_addr).await?;
+        assert_eq!(target_addr, socket.peer_addr().unwrap());
+        debug!("Client connected to validation server");
+        socket
+    };
+
+    tokio::select! {
+        is_readable = stream.readable() => {
+            is_readable.context("cannot read off client socket")?;
+        }
+
+        () = tokio::time::sleep(timeout) => {
+            anyhow::bail!("timed-out ({:?}) waiting for socket to become readable", timeout);
+        }
+    };
+
+    let mut buf = [0u8; 100];
+    let read_sz = stream.read(&mut buf[..REDIRECT_RESPONSE.len()]).await?;
+    let resp = String::from_utf8(buf[..REDIRECT_RESPONSE.len()].to_vec())?;
+    debug!(redirect_response = %resp, bytes_read = %read_sz);
+    anyhow::ensure!(
+        resp == REDIRECT_RESPONSE,
+        "expected client to receive {:?}, got {:?} instead",
+        REDIRECT_RESPONSE,
+        resp
+    );
+    Ok(())
+}
+
+#[tracing::instrument(name = "outbound_server", skip(ready_tx, shutdown))]
+async fn outbound_serve(
+    listen_addr: SocketAddr,
+    ready_tx: oneshot::Sender<Result<()>>,
+    shutdown: kubert::shutdown::Watch,
+) -> Result<()> {
+    let listener = TcpListener::bind(listen_addr)
+        .await
+        .expect("Failed to bind server");
+    info!("Listening for incoming connections");
+
+    if ready_tx.send(Ok(())).is_err() {
+        error!("Failed to send 'ready' signal, receiver dropped");
+        anyhow::bail!("Failed to bind server");
+    }
+
+    let resp_bytes = REDIRECT_RESPONSE.as_bytes();
+    tokio::spawn(accept(listener, resp_bytes));
+
+    let _handle = shutdown.signaled().await;
+    debug!("Received shutdown signal");
+    Ok(())
+}
+
+#[tracing::instrument(name = "outbound_accept", skip_all)]
+async fn accept(listener: TcpListener, resp_bytes: &'static [u8]) {
+    loop {
+        let (mut stream, client_addr) = listener
+            .accept()
+            .await
+            .expect("Failed to establish connection");
+        info!("Accepted connection");
+        let _ = tokio::spawn(async move {
+            match stream.write_all(resp_bytes).await {
```

Comment:
Generally we'd want to enforce some timeout to prevent leaking tasks as clients connect. In practice, this probably isn't necessary as these writes should basically always succeed, so we can just leave a comment documenting that.

```suggestion
            // We expect this write to complete instantaneously, s
            // a timeout is not needed here.
            match stream.write_all(resp_bytes).await {
```

Reply:
Gotcha. So, just to see if I understand correctly, under normal conditions, we'd want a timeout here to make sure the task is dropped if the write hangs?

Reply:
yeah. imagine a case where a client makes thousands of connections to the server and never reads on the sockets. the server could bloat and would eventually OOM.

I don't think it's particularly concerning here, but pointing this out for completeness.